src/GafferImage/ImageWriter.cpp

```diff
@@ -66,13 +70,438 @@ static InternedString g_compressionQualityPlugName( "compressionQuality" );
+		// data window. The origin of each tile is still the bottom left, and
+		// the tile index is calculated from the top left tile, incrementing in
+		// x first.
+
+		inline size_t outTileIndex( const Imath::V2i &tileOrigin ) const
+		{
+			return ( ( ( m_outputDataWindow.max.y - m_spec.tile_height - tileOrigin.y ) / m_spec.tile_height ) * m_numTiles.x ) + ( ( tileOrigin.x - m_outputDataWindow.min.x ) / m_spec.tile_width );
+		}
+
+		inline Imath::V2i outTileOrigin( const size_t tileIndex ) const
+		{
+			return Imath::V2i( ( ( tileIndex % m_numTiles.x ) * m_spec.tile_width ) + m_outputDataWindow.min.x, m_outputDataWindow.max.y - m_spec.tile_height - ( ( tileIndex / m_numTiles.x ) * m_spec.tile_height ) );
+		}
+
+		inline Imath::V2i outTileOrigin( const Imath::V2i &point ) const
+		{
+			Imath::V2i tempPoint = point - Imath::V2i( m_outputDataWindow.min.x, m_outputDataWindow.max.y );
+			Imath::V2i tileOrigin;
+			tileOrigin.x = tempPoint.x < 0 && tempPoint.x % m_spec.tile_width != 0 ? ( tempPoint.x / m_spec.tile_width - 1 ) * m_spec.tile_width : ( tempPoint.x / m_spec.tile_width ) * m_spec.tile_width;
+			tileOrigin.y = tempPoint.y < 0 && tempPoint.y % m_spec.tile_height != 0 ? ( tempPoint.y / m_spec.tile_height - 1 ) * m_spec.tile_height : ( tempPoint.y / m_spec.tile_height ) * m_spec.tile_height;
+			return tileOrigin + Imath::V2i( m_outputDataWindow.min.x, m_outputDataWindow.max.y );
+		}
+
+		inline Imath::Box2i outTileBounds( const Imath::V2i &point ) const
+		{
+			Imath::V2i origin = outTileOrigin( point );
+			return Imath::Box2i( origin, origin + Imath::V2i( m_spec.tile_width, m_spec.tile_height ) );
+		}
+
+		inline Imath::Box2i outTilesBounds() const
+		{
+			return Imath::Box2i( outTileOrigin( m_outputDataWindow.min ), outTileOrigin( m_outputDataWindow.max - Imath::V2i( 1 ) ) + Imath::V2i( m_spec.tile_width, m_spec.tile_height) );
+		}
+
+		inline bool firstChannelOfTile( const size_t channelIndex )
+		{
+			return channelIndex == 0;
+		}
+
+		inline bool lastChannelOfTile( const size_t channelIndex )
+		{
+			return channelIndex == ( m_spec.channelnames.size() - 1 );
+		}
+
+		void flagFilledTiles( const Imath::Box2i &inTileBounds )
+		{
+			for( size_t i = m_nextTileIndex; i < m_tilesData.size(); ++i )
+			{
+				if( !m_tilesFilled[i] )
+				{
+					Imath::Box2i outTileBnds( outTileBounds( outTileOrigin( i ) ) );
+					if( inTileBounds.max.x >= outTileBnds.max.x && inTileBounds.min.y <= outTileBnds.min.y )
+					{
+						m_tilesFilled[i] = true;
+					}
+					else
+					{
+						break;
+					}
+				}
+			}
+		}
+
+		void writeFilledTiles()
+		{
+			size_t tileIndex;
+			for( tileIndex = m_nextTileIndex; tileIndex < m_tilesData.size(); ++tileIndex )
+			{
+				Imath::V2i tileOrigin = outTileOrigin( tileIndex );
+
+				if( m_tilesFilled[tileIndex] )
+				{
+					writeTile( tileOrigin, m_tilesData[tileIndex] );
+					m_tilesData[tileIndex].reset();
+				}
+				else
```

Comment:
Any reason this isn't an `else if` directly? I found the internal if/break logic confusing enough that I went back up and re-read the whole explanation comment to make sure I understood what was happening.